### ISSUE TITLE
Transparency on screenshot

### ISSUE THREAD

**Issue by Czaki** [napari/napari#1566]
## 🐛 Bug

When create screenshot using Save Screenshot it contains transparency channel. When such screenshot is opened in programs which mark transparency (like Eye of Gnome or GIMP) it looks strange. 

![Zrzut ekranu z 2020-08-21 01-50-22](https://user-images.githubusercontent.com/3826210/90836833-b910db80-e350-11ea-92d6-b68ca2f5b600.png)
instead of

![Zrzut ekranu z 2020-08-21 01-40-15](https://user-images.githubusercontent.com/3826210/90836481-d09b9480-e34f-11ea-8b08-6727d7e7326e.png)

## To Reproduce

1. load image
2. add labelling
3. set opacity below 1
4. take screenshot 

## Expected behavior

Possibility of save screenshot without transparency, like done with print screen. 

## Environment

 - Please copy and paste the information at napari info option in help menubar here:
napari: 0.3.7.dev37+g0f5451a
Platform: Linux-5.4.0-42-generic-x86_64-with-debian-bullseye-sid
Python: 3.7.7 (default, Apr 20 2020, 16:13:22) [GCC 9.2.1 20191008]
Qt: 5.14.2
PyQt5: 5.14.2
NumPy: 1.19.1
SciPy: 1.5.1
Dask: 2.21.0
VisPy: 0.6.4

GL version: 4.6.0 NVIDIA 390.138
MAX_TEXTURE_SIZE: 32768

Plugins:
- PartSeg: 0.11.3.dev1+gc4e59a1.d20200817
- napari-plugin-engine: 0.1.6
- svg: 0.1.3
 - Any other relevant information:

**Comment by sofroniewn** [comment]
Hmm that does look funny, and definitely like a bug that we should fix. I wonder if it relates to #673 - we'll need to look in more. Thanks for flagging!!

**Comment by Czaki** [comment]
I check this on Ubuntu 18.04, Ubuntu 20.04, MacOS 10.14.6. In Macos default image viewer it is invisible, but when transfer on Ubuntu and open in Eye of Gnome it is visible.

**Comment by sofroniewn** [comment]
This may be related to #1237 too

**Comment by sofroniewn** [comment]
Could also relate to https://github.com/vispy/vispy/issues/1889 which is not screenshot specific but pyqt5 specific

**Comment by Czaki** [comment]
![Zrzut ekranu z 2020-08-25 00-08-11](https://user-images.githubusercontent.com/3826210/91101329-19608f80-e667-11ea-9619-8fb180f34fff.png)
Same for PySide2.

But #1237 looks similar.  Maybe @jni could check if removing alpha channel will help.

**Comment by kne42** [comment]
Hi @Czaki, thank you so much for taking the time to make this bug report!

I was able to reproduce this bug. My preliminary investigation leads me to believe that this is likely to be an error in vispy's canvas rendering code. I'll take a look and see if I can find any solutions.

**Comment by djhoese** [comment]
Hi everyone. A couple thoughts coming from someone who doesn't know the inner workings of napari but does know vispy:

1. How is napari making the screenshots? Does it use the `_screenshot` method in vispy?
2. What does this look like in napari? If the regions aren't 100% opaque in napari then wouldn't we expect to see the transparency (and the image viewer's checkerboard background)? If I'm understanding the original bug report, we're expecting to see solid colors in the screenshot?

**Comment by Czaki** [comment]
@djhoese 
It looks like it is not `_screenshot` but `grabFramebuffer` 
https://github.com/napari/napari/blob/2ff87e56aac2e1c812e82fa5bbf8bcf1b7954150/napari/_qt/qt_viewer.py#L365

here screenshot of all window 
https://github.com/napari/napari/blob/2ff87e56aac2e1c812e82fa5bbf8bcf1b7954150/napari/_qt/qt_main_window.py#L509

I expect that if any layer has opacity 100% then there should be no opacity on screenshot. 
But I do not understand why there is decision to kep alpha channel on screenshot (which is intentional)
https://github.com/napari/napari/blob/2ff87e56aac2e1c812e82fa5bbf8bcf1b7954150/napari/_qt/qt_viewer.py#L377

Maybe problem is that screenshot method does not grab information from background?

**Comment by zeroth** [comment]
@djhoese 

I am not 100% sure, but in my preliminary finding I found that
`self.canvas.native` is of type `vispy.app.backends._qt.CanvasBackendDesktop`
and looking at `vispy` code CanvasBackendDesktop inherits QGLWidget.

QGLWidget has a method grabFrameBuffer which has a default paramater `withAlpha = false`  (`QImage QGLWidget::grabFrameBuffer(bool withAlpha = false)`) 

I think before digging further we should test it with `withAlpha = true`.

https://doc.qt.io/qt-5/qglwidget.html#grabFrameBuffer

**Comment by djhoese** [comment]
Note the misleading import in `_qt.py`:

https://github.com/vispy/vispy/blob/bb2cd96eb16dde1ecac3c70f9e572791bb5699e9/vispy/app/backends/_qt.py#L82

It is actually the newer QOpenGLWidget.

@Czaki You said:

> In Macos default image viewer it is invisible, but when transfer on Ubuntu and open in Eye of Gnome it is visible.

When you say "transfer on Ubuntu", do you mean you ran the same napari operations on Ubuntu, saved an image, and then viewed it? Or do you mean you copied the screenshot image to an Ubuntu machine and viewed it?

It is hard for me to tell from previous comments if this is OSX only or not. If it is OSX only then I strongly recommend looking at https://github.com/vispy/vispy/issues/1889 which @sofroniewn already mentioned. Also, iirc napari is able to use PyQt5 *or* PySide2, right? If so, is this an issue with PySide2 or only PyQt5?

**Comment by zeroth** [comment]
@djhoese Thanks for pointing that. I found this bug report on Qt https://bugreports.qt.io/browse/QTBUG-55245 might give you some idea.

**Comment by djhoese** [comment]
Hm the last comment on that bug report suggests that should be fixed with Qt 5.12 which I think conda-forge has as the newest version. I wonder if this was a bug back on Qt 5.9. Might be hard to check.

**Comment by Czaki** [comment]
> When you say "transfer on Ubuntu", do you mean you ran the same napari operations on Ubuntu, saved an image, and then viewed it? Or do you mean you copied the screenshot image to an Ubuntu machine and viewed it?

I mean copy screnshoot image on ubuntu machine and there view it with build in viewer (Eye of Gnome). 

I created screenshot which have this effect on Ubuntu 18.04, 20.04 using PyQt 5.14.2, on Ubuntu 20.04 using PySide2 5.15.0,   on MacOs using PyQt  5.14.2. I do not test other version of Qt bindings. I use Qt bindings from PyPi, not conda one.

**Comment by djhoese** [comment]
I think I need this put another way. What exactly is the unexpected behavior? I'm not sure I understand what moving a screenshot from one OS to another tells us besides how different image viewers display transparent pixels. One image viewer might use a black background or a white background so transparent pixels always have that behind them. Others might use a checkerboard pattern to make it clear that there is some transparency. Usually this is configurable in the viewer.

In napari, are these extra "shapes" on the image additional napari layers on top of the background image? Or is this displaying one single image layer where these pixels are being "masked out"?

In the extra layer case, I would consider it a bug if the "shapes" had some level of transparency that inside napari allowed the background image to show through, but when saved to a screenshot they are "cutting through" the background image to show transparency all the way through.

In the single image layer case, this looks expected to me if the "shapes" have some level of transparency but napari is setting the GL background (the GL clear color) to black or white or some other color (I think you have to do this) so it doesn't look fully transparent. I suppose this is a bug in the sense that Qt isn't using the GL clear color as part of the screenshot but I'm not sure that is necessarily defined.

**Comment by Czaki** [comment]
> What exactly is the unexpected behavior?
Below labels layer there is image layer. This layer is not transparent. So on screenshot I do not expect transparency in this place. 
I expect transparency on screenshot only if all layers in stack have set opacity below 1

> I'm not sure I understand what moving a screenshot from one OS to another tells us besides how different image viewers display transparent pixels

Moving is only to use viewer which marks pixels with alpha value bellow 255. I do not know such viewer for MacOS but I expect that it exists.

**Comment by djhoese** [comment]
On OSX, Preview can be configured to show a different background (checkerboard even I think).

**Comment by kne42** [comment]
Hi @djhoese, thanks for taking a look!

To try to clear things up:
- these are effectively two ImageVisuals
- the one closer to us ("the foreground") is partially transparent
- the one farther from us ("the background") is completely opaque
- when we take a screenshot, we expect that because the background overlaps with the foreground, the result should be a blending of the two, but opaque, since the background is opaque
- instead we wind up with patches of transparency in places where the two overlap

[Here is a notebook](https://gist.github.com/kne42/ceb41bca2b68f85a350349b7da523d8c) that may help you reproduce the results. You can see that there is transparency in the array that gets returned back to us. There's also some weird color inversion in there that you can ignore since I believe it's either Qt or napari's fault.

Some weird thing that I don't understand though is that the arrays returned by `canvas.native.grabFrameBuffer()` (once converted to numpy) and `canvas.render()` are different.

**Comment by djhoese** [comment]
Are you using `canvas.render()` to do the whole screenshot? I know it was said that the `grabFrameBuffer()` was being used above, but is it being used for both cases with and without the canvas? Have you tried using `vispy.gloo.util._screenshot` (yes I know its private, sorry). Here is what I got from your notebook on my PopOS (Ubuntu) system (using matplotlib's `Figure.savefig`):

The actual napari window screenshotted with a separate tool:

![image](https://user-images.githubusercontent.com/1828519/93116388-d2245800-f682-11ea-9d8f-73e817168c3a.png)

Your `ss_w_viewer`:

![ss_w_viewer](https://user-images.githubusercontent.com/1828519/93115989-4ad6e480-f682-11ea-8e72-422f5aec8985.png)

^ This shows the weird inverted colors you were talking about. Your `ss_wo_viewer`:

![ss_wo_viewer](https://user-images.githubusercontent.com/1828519/93116009-532f1f80-f682-11ea-9f9d-9efaa71979a0.png)

^ This looks right-ish here on GitHub because it is adding a background (white from the web page?) to it I think. On my local image viewer the alpha applies to the whole thing and I can see the checkerboard background:

![image](https://user-images.githubusercontent.com/1828519/93116241-a2755000-f682-11ea-9bb4-b316532e0ffb.png)

I then added some code to the end of your notebook:

```
from vispy.gloo.util import _screenshot
ss_vispy = _screenshot()
f = plt.figure()
plt.imshow(ss_vispy)
plt.axis('off');
# ---
f.savefig("/tmp/ss_vispy.png")
```

![ss_vispy](https://user-images.githubusercontent.com/1828519/93116018-56c2a680-f682-11ea-8e33-9d65f20f682e.png)

Unlike the `ss_wo_viewer` one above, this one seems to include the background (or rather take away the transparency). A screenshot of the image viewer on my system:

![image](https://user-images.githubusercontent.com/1828519/93117591-b02bd500-f684-11ea-887d-23b7512ca9a9.png)

Checking the `ss_vispy` array I see that the alpha is all 255s:

```
(ss_vispy[:, :, 3] == 255).all()
# True
```

I skimmed the source code as I was working on this comment and I'm guessing you aren't ever using `.render()` in napari? In general I've been suggesting people use the `_screenshot` utility as the results seem to be more expected. I hope to make it a public function in the next vispy release (PRs welcome while I work on other maintenance stuff).

**Comment by djhoese** [comment]
FYI A fix for this should be released as part of VisPy 0.7 (hopefully this week). It was fixed in https://github.com/vispy/vispy/pull/2090... we think.
